test/test_adjoint.py

```diff
@@ -167,3 +158,60 @@ def test_optimized_plated_einsum_adjoint(equation, plates, backend):
         assert isinstance(actual, funsor.Tensor)
         assert expected.shape == actual.data.shape
         assert torch.allclose(expected, actual.data, atol=1e-7)
+
+
+@pytest.mark.parametrize('num_steps', list(range(3, 13)))
+@pytest.mark.parametrize('sum_op,prod_op,state_domain', [
+    (ops.add, ops.mul, bint(2)),
+    (ops.add, ops.mul, bint(3)),
+    (ops.logaddexp, ops.add, bint(2)),
+    (ops.logaddexp, ops.add, bint(3)),
+], ids=str)
+@pytest.mark.parametrize('batch_inputs', [
+    {},
+    {"foo": bint(5)},
+    {"foo": bint(2), "bar": bint(4)},
+], ids=lambda d: ",".join(d.keys()))
+@pytest.mark.parametrize('impl', [
+    sequential_sum_product,
+    naive_sequential_sum_product,
+    xfail_param(MarkovProduct, reason="something missing?"),
```

Comment:
These test cases fail mysteriously at the final assertion with incorrect tensor values

Reply:
My first guess is an alpha renaming issue, which is the main difference between `sequential_sum_product` and `MarkovProduct`.

Reply:
In particular, the keys and values of the `MarkovProduct.step_names` dict will typically unmangle to the same name.